initial tests for globals

marianoguerra opened the pull request:
this are some simple tests to check that globals are set and return the set value both from inside the module and from outside and that setting two different globals doesn't affect each other.

this is not in sync with the spec which says that store should return the stored value, will look into it now.

lukewagner commented:
Thanks!  I think you also need to add a rule to check.ml to allow expressions like assigning a `set_global` to a local to succeed.  Just a fair warning, we're seriously considering removing globals from the MVP (and so ml-proto) in #154.  However the analogous change you've made here also applies to `set_local` and `store`.

marianoguerra commented:
* added the change in check for store_global
* added test to check that I can use the result of store_global
* added the ability to return the value being set on set_local
* added test to check that I can use the result of set_local

regarding removing globals, no problem, I'm using this as a training to understand the project and the spec so no problem for me throwing this later :). Just out of curiosity, will globals be removed for good or there are plans to reintroduce them later? I think globals are easily "shimed" with module memory so maybe they are not that primitive...

lukewagner commented:
@marianoguerra Globals would just be removed from the MVP.  See the new #344 (which, when merged, will be followed by a PR in this repo) for more explanation.

lukewagner commented on ml-proto/src/spec/check.ml:
```diff
@@ -179,14 +179,14 @@ let rec check_expr c ts e =
   | SetLocal (x, e1) ->
     check_expr c [local c x] e1;
-    check_type [] ts e.at
+    check_type [local c x] ts e.at
```
Here and below, it might be nice to factor out via `let t1 = local c x in` to avoid duplicating the expression.

marianoguerra replied:
I will, just need to read an introduction to ocaml somewhere, any good recommendation? (I already code in clojure, erlang and scheme so it can be a high level introduction)

rossberg-chromium replied:
FWIW, I actually find it more readable (since shorter) as is. :)

lukewagner replied:
Hehe, fair enough.  I was on the edge anyhow for such a simple duplicated expr.

jfbastien commented:
Should we avoid merging this, based on https://github.com/WebAssembly/design/pull/344 removing globals from MVP?

marianoguerra commented:
@jfbastien I can remove globals and merge the other changes if it's ok for you.

jfbastien commented:
Sounds good.

lukewagner commented:
If you update this PR to remove globals, I think we can merge it.

lukewagner commented:
Created #106 which subsumes this PR.
